src/librustc_typeck/astconv.rs

```diff
@@ -1277,8 +1277,18 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx>+'o {
         // Find the type of the associated item, and the trait where the associated
         // item is declared.
         let bound = match (&ty.sty, ty_path_def) {
+            (&ty::Adt(adt_def, _), Def::TyAlias(_)) if adt_def.is_enum() => {
+                // Variant of enum resolved as type alias
+                let variant_def = adt_def.variants.iter().find(|vd| vd.name == assoc_name.name);
```

Comment:
This is... well, hygiene is not done correctly.
Look at any use of `hygienic_eq` in rustc_typeck for examples of how to find the correct associated item hygienically.
Except that HIR for variants doesn't have relevant data, so `VariantKind::name` needs to be replaced with `VariantKind::ident`.
That's all about macros 2.0 though, so not critical.

Reply:
Makes sense, ta.